### ISSUE TITLE
search_is_some fixit hint suggests invalid code

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#4033]
`clippy 0.0.212 (9897442 2019-04-23)`

in this code
````rust
    // https://docs.rs/walkdir/2.2.7/walkdir/#example-skip-hidden-files-and-directories-on-unix
    fn is_git(entry: &walkdir::DirEntry) -> bool {
        entry
            .path()
            .components()
            .find(|&path_elm| path_elm == std::path::Component::Normal(OsStr::new(".git")))
            .is_some()
    }
````
clippy suggests:
````
warning: called `is_some()` after searching an `Iterator` with find. This is more succinctly expressed by calling `any()`.
   --> src/main.rs:352:9
    |
352 | /         entry
353 | |             .path()
354 | |             .components()
355 | |             .find(|&path_elm| path_elm == std::path::Component::Normal(OsStr::new(".git")))
356 | |             .is_some()
    | |______________________^
    |
    = note: #[warn(clippy::search_is_some)] on by default
    = note: replace `find(|&path_elm| path_elm == std::path::Component::Normal(OsStr::new(".git"))).is_some()` with `any(|&path_elm| path_elm == std::path::Component::Normal(OsStr::new(".git")))`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#search_is_some
````

if I apply the suggestion
````rust
    fn is_git(entry: &walkdir::DirEntry) -> bool {
        entry
            .path()
            .components()
            .any(|&path_elm| path_elm == std::path::Component::Normal(OsStr::new(".git")))
    }
````
the code no longer compiles
````
    Checking raca v0.1.0 (/home/matthias/vcs/github/raca)
error[E0308]: mismatched types
   --> src/main.rs:355:19
    |
355 |             .any(|&path_elm| path_elm == std::path::Component::Normal(OsStr::new(".git")))
    |                   ^^^^^^^^^ expected enum `std::path::Component`, found reference
    |
    = note: expected type `std::path::Component<'_>`
               found type `&_`
    = help: did you mean `path_elm: &std::path::Component<'_>`?

error: aborting due to previous error
````

The correct code is
````rust
.any(|path_elm| ...
````
instead of
````rust
.any(|&path_elm| ...
````

**Comment by flip1995** [comment]
FnDecl of `any`:
```rust
fn any<F>(&mut self, f: F) -> bool 
    where F: FnMut(Self::Item) -> bool,
```

FnDecl of `find`:
```rust
fn find<P>(&mut self, predicate: P) -> Option<Self::Item> 
    where P: FnMut(&Self::Item) -> bool, 
```

The closure of `find`takes the element by ref, while the closure of `any` takes the element by value. So this should be easily fixable by checking if we match on `|&_|` in `find` and if so only use `|_|` in the suggestion.
